[v1] Fix placeholder

seaneking opened the pull request:
The placeholder for simpla-img is currently pretty broken. Turned out a lot of this was due to us working with an image with an empty `src`, which is actually not allowed and handled differently by every browser. 

This PR shims these issues by making the 'empty' state of `src` a 1px transparent PNG (as data URI). The transparent PNG is 76 bytes which I think is acceptable (couldn't use GIF, because cropper spits back a PNG instead - this might be patched in `1.0-gif` branch, if so we should use GIF pixel because it's about half the size at 48 bytes).

Also makes placeholder configurable with `placeholder` attribute (can take any valid CSS background prop - eg: colors, img urls).

Tested in Chrome, Safari, Firefox, and Edge.

Fixes https://github.com/SimplaElements/simpla-img/issues/47 and https://github.com/SimplaElements/simpla-img/issues/46 (though still kinda fugly that empty editor opens when placeholder clicked, but no longer breaking bug).

Oh and renamed `editing` to `active` to be consistent with other elements while I was in there.

seaneking commented on src/behaviors/constants.js:
```diff
@@ -0,0 +1,7 @@
+export const EDITOR_COMPONENT = 'simpla-img-editor.html';
```
I realise it's kind of random having these in here and not others - got halfway through moving all constants to this file and realised that might not be the best idea anyway. Wasn't sure where to put `DEFAULT_SRC` other than here, and seemed a bit daft having a constants file with one constant.

What say you @bedeoverend - move all constants here, or move `DEFAULT_SRC` elsewhere?

seaneking commented on src/simpla-img.html:
```diff
@@ -100,35 +108,36 @@
-       * @param  {Boolean} editable Value of the editable prop
-       * @param  {Boolean} editing Value of the editable prop
-       * @param  {String} src       Value of the src prop
+       * @param  {Boolean} placeholder String to apply to background
```
Random aside - I've started being more descriptive in my JSDoc comments ('Value of the editable prop' isn't very helpful), going to try and follow this pattern of what the arg is actually used for in the function/by the user. Otherwise they end up like this https://twitter.com/randal_olson/status/843831671648669696

bedeoverend replied:
Haha, yeah I've been reading some really interesting stuff on code comments - basically like, they're a necessary evil. You should only use them if your code doesn't make sense on it's own, and you should try your hardest to make your code make sense on it's own. Regardless though, JSDoc ones are good for exporting to docs (eventually)

seaneking commented:
Okay thought 'hey should make this pass tests' but then realised tests were totes broken on Travis and got a bit carried away. Have a gulp + travis refactor that should fix them sitting locally, but will push to 1.0 when this is merged. Also all tests pass locally

bedeoverend commented on gulpfile.babel.js:
```diff
@@ -77,74 +79,69 @@ const bs = browserSync.create(),
 
 gulp.task('watch:src', () => gulp.watch(['src/**/*'], () => gulprun('build', 'refresh')));
-gulp.task('watch:tests', () => gulp.watch(['src/**/*', 'test/**/*'], () => gulprun('build:tests')))
+gulp.task('watch:tests', () => gulp.watch(['test/**/*'], () => gulprun('build:tests')))
```
This needs to watch the `src` as the `tests` pull in exported functions, see the `viewable` unit tests.

seaneking replied:
oh sorry, thought it was a copy/paste typo or something

bedeoverend commented on gulpfile.babel.js:
```diff
@@ -77,74 +79,69 @@ const bs = browserSync.create(),
 gulp.task('watch', ['watch:src', 'watch:tests']);
 
-gulp.task('default', ['build', 'build:tests', 'serve', 'watch']);
+gulp.task('serve', (callback) => bs.init(OPTIONS.browserSync));
```
FYI don't need callback here, unused param. Doesn't matter though

seaneking replied:
ah, good spot. From eons ago haha.

bedeoverend commented:
Tests are still shitty on `img` - want to write some nightwatch ones that are more generalised.

Also made review + comment

seaneking commented:
Updated with a check for default src value in Simpla connector (also fixed that gulpfile typo, oops).

Also, what's with the `window._counter` in simpla behavior? Leftover from debugging or something?

bedeoverend commented:
👍 

And yup, my bad 😞 was for debugging.

seaneking commented:
Ha figured, there were a few console logs lying around too (logging all the base64 data, was wondering why my inspector was freezing up haha). Will clean up and merge, cools?

bedeoverend commented:
LGTM 👍
